core/src/main/scala/org/apache/spark/sql/execution/CachedPlanHelperExec.scala

```diff
@@ -97,7 +98,19 @@ case class CachedPlanHelperExec(childPlan: CodegenSupport, @transient session: S
   override def doConsume(ctx: CodegenContext, input: Seq[ExprCode], row: ExprCode): String =
     parent.doConsume(ctx, input, row)
 
-  override protected def doExecute(): RDD[InternalRow] = childPlan.execute()
+  override protected def doExecute(): RDD[InternalRow] = {
+    // This does not handle nested aggregates or joins. Should we handle ?
+    val repeatSeq = childPlan.collect {
+      case p: NonRecursivePlans => !p.producedForPlanInstance
```

Comment:
Why do we need methods/vars in NonRecursivePlans? Any Node that implements NonRecursivePlans we, return false. What is the use of calling startProducing at the beginning of doProduce when we know that producedForPlanInstance is always be going to be true for this node.